src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala

```diff
@@ -235,6 +238,61 @@ class UserService(protected val userInfo: UserInfo, val dataSource: SlickDataSou
     }
   }
 
+  private def verifyNoSubmissions(userRef: RawlsUserRef, dataAccess: DataAccess): ReadAction[PerRequestMessage] = {
+    dataAccess.submissionQuery.findBySubmitter(userRef.userSubjectId.value).exists.result flatMap {
+      case true => DBIO.failed(new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, "Cannot delete a user with submissions")))
+      case _ => DBIO.successful(RequestComplete(StatusCodes.OK))
+    }
+  }
+
+  // can't just delete DB rows; also removes user from Google Groups
+  private def removeUserFromAllGroups(userRef: RawlsUserRef, dataAccess: DataAccess): ReadWriteAction[RawlsUser] = {
+    withUser(userRef, dataAccess) { user =>
+      dataAccess.rawlsGroupQuery.listGroupsForUser(userRef) flatMap { groupRefs =>
+        // Slick complains if it remains a Set
+        val removeActions = groupRefs.toSeq.map { groupRef =>
+          withGroup(groupRef, dataAccess) { group =>
+            updateGroupMembersInternal(group, Set(user), Set.empty, RemoveGroupMembersOp, dataAccess)
```

Comment:
this is certainly correct but potentially slow as this makes repeated google calls to remove the user from groups plus database stuff all per group.

Instead it probably could be a single delete query: delete from GROUP_USERS where .... (done via slick of course). I also don't think we need to worry about removing the user from groups because deleting the user's proxy group will probably take care of that.